feat(jellyfish-api-core): Add updatemasternode RPC typings

marktanrj opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:

- Add RPC updatemasternode

#### Which issue(s) does this PR fixes?:
<!--
(Optional) Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes part of https://github.com/JellyfishSDK/jellyfish/issues/1842

marktanrj committed:
Add updateMasternode rpc (WIP)

marktanrj committed:
Add test cases

marktanrj committed:
Undo update config that was accidentally committed

marktanrj committed:
Undo update config that was accidentally committed

marktanrj committed:
Update testcases

marktanrj committed:
Merge branch 'main' into marktanrj/updatemasternode-rpc-typing

marktanrj committed:
Update image version

marktanrj committed:
Update image version

marktanrj committed:
Update image version

shohamc1 committed:
feat(v3): bump to grandcentral and use `master-1d169a751` (#1841)

Bump testcontainers to `grandcentral` with `master-1d169a75`. Fixes all broken tests and known regression from this change.

Co-authored-by: Dilshan Madushanka <helloscoopz@gmail.com>

shohamc1 committed:
chore(jellyfish-api-core): Change assert order in `getMiningInfo.test.ts` to match other tests (#1865)

<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:
Uses to the assertion order used by all other tests. Uses a regex
pattern to match instead of `toContain`.

#### Additional comments?:

As requested in
https://github.com/JellyfishSDK/jellyfish/pull/1841#discussion_r1028115217

marktanrj committed:
Update image version

marktanrj committed:
Update tests

shohamc1 committed:
feat(jellyfish-api-core): defid v3 poolpair/poolswap internal changes (#1854)

<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:

Add tests for new fixes on defid v3.

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,276 @@
+import { AddressType } from '../../../src/category/wallet'
+import { RpcApiError } from '@defichain/jellyfish-api-core'
+
+describe('Masternode', () => {
```
Perhaps use a more elaborate title here.

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,380 @@
+import { MasterNodeRegTestContainer } from '@defichain/testcontainers/dist/index'
```
```suggestion
import { MasterNodeRegTestContainer } from '@defichain/testcontainers'
```

marktanrj replied:
Thanks, amended

infinia-yzl reviewed (changes requested):
Please also cover the following test scenarios:
- Check new state is TRANSFERRING and owner is still the same
- Test update while masternode is not active
- Test update while masternode is in TRANSFERRING state
- Test PRE_ENABLED state and owner change
- Test update during PRE_ENABLED state
- Test ENABLED state and owner change
- Simultaneous owner / operator / reward updates
- Test empty argument
- Test owner update without collateral input
- Test incorrect new collateral amount
- Try updating two nodes to the same address
- Test updating several MNs owners in the same block
- Node 1 try to update node 0 which should be rejected

Might be more to add, please refer to https://github.com/DeFiCh/ain/blob/dfe240ceb8247d36202e21e9ae5034d18a06eeba/test/functional/rpc_updatemasternode.py.

shohamc1 committed:
fix(jellyfish-api-core): update testPoolSwap arguments (#1856)

<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:

Exposes the path and verbose arguments on `testpoolswap`. Removes the
defunct UTXO argument.

marktanrj committed:
Undo rebase mess up

marktanrj committed:
Merge main

marktanrj committed:
Add more tests

infinia-yzl commented:
@marktanrj please change PR description to "fixes part of", otherwise it will close the entire issue.

helloscoopa reviewed (changes requested):
Seems like still few tests are missing? Ticked found tests.

- [x] Check new state is TRANSFERRING and owner is still the same
- [x] Test update while masternode is not active
- [x] Test update while masternode is in TRANSFERRING state
- [x] Test PRE_ENABLED state and owner change
- [x] Test update during PRE_ENABLED state
- [x] Test ENABLED state and owner change
- [x] Simultaneous owner / operator / reward updates
- [x] Test empty argument
- [ ] Test owner update without collateral input - Test in DfTx
- [x] Test incorrect new collateral amount
- [x] Try updating two nodes to the same address
- [x] Test updating several MNs owners in the same block
- [x] Node 1 try to update node 0 which should be rejected — hmm I think for this we need to setup a test group?

helloscoopa commented on packages/jellyfish-api-core/src/category/masternode.ts:
```diff
@@ -144,6 +144,24 @@ export class Masternode {
+   * @param {UTXO[]} [utxos = []] Array of specified utxos to spend.
+   * @param {string} [utxos.txid] The transaction id.
+   * @param {number} [utxos.vout] The output number.
+   * @return {Promise<string>} Resignation Transaction.
```
Minor:

```suggestion
   * @return {Promise<string>} The hex-encoded hash of broadcasted transaction.
```

or anything else that is meaningful.

marktanrj committed:
Update packages/jellyfish-api-core/src/category/masternode.ts

Co-authored-by: Dilshan Madushanka <helloscoopz@gmail.com>
Signed-off-by: Mark Tan <35588098+marktanrj@users.noreply.github.com>

imkven committed:
Test update MN during transferring

shohamc1 commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,492 @@
+        message: expect.stringContaining(`Masternode ${masternodeId} is not in 'ENABLED' state`),
+        method: 'updatemasternode'
+      })
+    }
```
```suggestion
    const promise = client.masternode.updateMasternode(masternodeId, {
        ownerAddress: anotherNewAddress
    })
    await expect(promise).rejects.toThrow(RpcApiError)
    await expect(promise).rejects.toThrow(`Masternode ${masternodeId} is not in 'ENABLED' state`)
```

shohamc1 reviewed (changes requested):
@imkven

imkven committed:
Test updating several MNs owners in the same block

shohamc1 committed:
Refactor: remove try-catch blocks

shohamc1 committed:
Refactor: do not expect `undefined`

imkven committed:
Update updatemasternode.test.ts

imkven committed:
Add test: Node 1 try to update node 0 which should be rejected

imkven committed:
Update updatemasternode.test.ts

imkven committed:
Update updatemasternode.test.ts

imkven committed:
Update updatemasternode.test.ts

imkven committed:
Update updatemasternode.test.ts

imkven commented:
I have referred to `/test/functional/rpc_updatemasternode.py` to write tests in `updatemasternode.test.ts` but I have 2 tests that do not get the error message in Python. Can you help check? @shohamc1 CC @marktanrj @infinia-yzl 

    ○ skipped Test owner update without collateral input
    ○ skipped Node 1 try to update node 0 which should be rejected

shohamc1 committed:
Fix incorrect auth test

shohamc1 commented:
> ○ skipped Test owner update without collateral input

This test requires manually building an incorrect transaction, I think it is better suited for the tests on https://github.com/JellyfishSDK/jellyfish/pull/1863

> ○ skipped Node 1 try to update node 0 which should be rejected

Fixed in [ca6ca20](https://github.com/JellyfishSDK/jellyfish/pull/1847/commits/ca6ca2067fb3ca0424238851ba1aba28f997a312), but we should refactor the tests in the PR to use the `TestingGroup` on all tests.

imkven committed:
Merge branch 'main' into marktanrj/updatemasternode-rpc-typing

imkven committed:
Update updatemasternode.test.ts

imkven committed:
Merge branch 'main' into marktanrj/updatemasternode-rpc-typing

shohamc1 reviewed (dismissed):
:shipit:

fuxingloh commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,576 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
Can we use a deterministic state here?

imkven replied:
Added tests. https://github.com/JellyfishSDK/jellyfish/pull/1847/commits/f4620aa0953105f0f1ec2b19fab856a3c10c8b71

infinia-yzl replied:
I think it can be fully deterministic @imkven.

imkven replied:
Added test: https://github.com/JellyfishSDK/jellyfish/pull/1847/commits/399b6131c37ff2acf33f96380f85d51e4623c4b2

fuxingloh commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,576 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
Same for this

imkven replied:
Added tests. https://github.com/JellyfishSDK/jellyfish/pull/1847/commits/399b6131c37ff2acf33f96380f85d51e4623c4b2

fuxingloh reviewed (changes requested):
LGTM, with 1 minor change for clarity

imkven committed:
Update updatemasternode.test.ts

imkven committed:
Merge branch 'main' into marktanrj/updatemasternode-rpc-typing

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,590 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
```suggestion
      state: 'ENABLED',
```

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,590 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
```suggestion
      state: 'ENABLED',
```

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,590 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
```suggestion
      state: 'ENABLED',
```

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,590 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
```suggestion
      state: 'ENABLED',
```

infinia-yzl commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,590 @@
+      resignHeight: expect.any(Number),
+      resignTx: expect.any(String),
+      collateralTx: expect.any(String),
+      state: expect.any(String),
```
```suggestion
      state: 'ENABLED',
```

infinia-yzl reviewed (changes requested):
State is fully deterministic.

imkven committed:
Update updatemasternode.test.ts

fuxingloh commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,576 @@
+
+  it('should updateMasternode with utxos', async () => {
+    const balance = await container.call('getbalance')
+    expect(balance >= 2).toBeTruthy()
```
Kinda weird to write it like this, it should be

```suggestion
    expect(balance).toBeGreaterThanOrEqual(2)
```

fuxingloh commented on packages/jellyfish-api-core/__tests__/category/masternode/updatemasternode.test.ts:
```diff
@@ -0,0 +1,576 @@
+
+    {
+      const masternodes = await client.masternode.listMasternodes()
+      expect(masternodes[masternodeId]).toBeTruthy()
+      expect(masternodes[masternodeId].state).toStrictEqual('TRANSFERRING')
+      expect(masternodes[masternodeId].ownerAuthAddress).toStrictEqual(initialAddress)
```
```suggestion
      expect(data).toStrictEqual(expect.objectContaining({
        state: 'TRANSFERRING',
        ownerAuthAddress: initialAddress
      }))
```

For better readability, it's better to write tests like this. Most of the tests written previously were written the previous "style" that we've since discouraged. We don't have to change it now but just a thought for the future.

fuxingloh reviewed (approved):
LGTM, with a few suggestions that we don't have to apply now. More than 2000 test cases, I believe, are written in the "old style" that has been discouraged now.
